docs/CoinContractUpgrades.md

```diff
@@ -0,0 +1,154 @@
+
+## Coin-v*.pact upgrades
+
+In order to upgrade the coin contract source code, the following are needed:
+
+- A pact version bump: obviously, if changes to Pact are involved, the pact source repository reference in the `cabal.project` file must be modified with a reference to the correct git tag or commit.
+- [Bless](https://pact-language.readthedocs.io/en/stable/pact-reference.html?highlight=bless#blessing-hashes) previous coin versions: this allows for Kadena, as an upstream provider, to phase in v(n+1) upgrades without breaking downstream code that depends on older versions of the coin contract.
+- Make code changes as needed.
+- Repl test files: the repl files should be labeled `coin-vN.repl` (for some N), and should be a copy-over of the previous `coin-v(N-1).repl`, with additional tests for modified code. When a pull request is made to the repository, someone should check the diff between the two repl files to make sure nothing has changed needlessly. We should try to keep test semantics as stable as possible unless there's a good reason.
+- Create a `load-coin-contract-v*.yaml` script in the immediate directory with the following:
+
+  ```yaml
+  codeFile: coin-v*.pact
+  nonce: coin-contract-v*
+  keyPairs: []
+  ```
+
+  This will be used during the `Ea` upgrade transaction generation process.
+
+## Updating Genesis + Upgrade transactions
+
+In order to generate upgrade or new genesis transactions, the `ea` tool must be run on an updated script with the new transaction information added to the list of generated transactions. The following steps must be taken in order to generate the new modules that will reference these transactions for use in `chainweb-node` Haskell source:
+
+- Update `Ea`: this is a four step process.
+  1. `tools/ea/Ea/Genesis.hs` must be updated to reference the load YAML script defined in the previous section for coin contract changes.
+  2. `tools/ea/Ea.hs` must be updated to call `genTxModule` on the new coin contract reference (see: `genCoinV4Payloads` etc.). Example:
+  ```haskell
+    genCoinV4Payloads :: IO ()
+    genCoinV4Payloads = genTxModule "CoinV4"
+      [ fungibleXChainV1
+      , coinContractV4
+      ]
+
+    genCoinV5Payloads :: IO ()
+    genCoinV5Payloads = genTxModule "CoinV5"
+      [ coinContractV5
+      ]
+  ```
+  3. the newly updated `genCoinV*Payloads` function must be added to the `main` function. Keep in mind that if any additional upgrades are present, the transactions are generated in sequence, therefore are order-sensitive. If new code has dependencies, then those dependencies need to occur earlier in the input list to `genTxModule`.
+  4. The `ea` tool must be run by issuing the following command:
```

Comment:
alternatively IIRC you can just go into the Ea in the repl and run main IIRC. :^)

Useful if you're already in the repl session fixing something.

Reply:
Done :)